### PR TITLE
Implement Accept-Language locale detection

### LINKED ISSUE #139: task: Implement locale_from_accept_language server utility in ars-i18n
**Wave 5a** | **Points: 2** | **Epic: #54 (I18n)**

## Goal

Implement server-side locale detection from HTTP Accept-Language header.

## Layer

Subsystem

## Framework

None

## Test tier

Unit

## Depends on

- #75 (ICU4X Locale integration)

## Spec refs

- `spec/foundation/04-internationalization.md` §10.3

## Scope

- `locale_from_accept_language(header: &str, supported: &[Locale]) -> Locale` behind `#[cfg(feature = "std")]`
- Quality value parsing ("en-US,en;q=0.9,de;q=0.8" format)
- Sort by quality descending
- Language-only fallback matching (header "pt-BR" matches supported "pt" if "pt-BR" not in list)
- Fallback to first supported locale, then en-US

## Tests to add first

- `"en-US,en;q=0.9,de;q=0.8"` with supported `[en_us, de]` returns `en_us`
- Language-only matching: header "pt-BR" matches supported "pt" if "pt-BR" not supported
- Quality ordering: lower-quality entries are tried after higher-quality
- Malformed quality values (NaN) handled gracefully
- Falls back to first supported locale when no match
- Falls back to en-US when supported list is empty
- Only available with `std` feature

## Acceptance criteria

- Function matches spec §10.3 exactly
- `#[cfg(feature = "std")]` gated
- Malformed input handled gracefully (no panics)
- All tests TDD-first, zero warnings

### PR DESCRIPTION
**Pull request by fogodev**
Closes #139

## Summary
- add `locale_from_accept_language` to `ars-i18n` behind the `std` feature gate
- parse `Accept-Language` preferences with quality sorting, exact locale matching, language-only fallback, and safe fallback behavior
- add focused tests for malformed quality values, invalid locale tags, and empty header entries

## Verification
- `cargo test -p ars-i18n detect --quiet`
- `cargo llvm-cov test -p ars-i18n --text -- detect`
- `cargo xci`

### CODE REVIEW

File: crates/ars-i18n/src/detect.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            Some((tag, quality))
+        })
+        .collect::<Vec<_>>();
+
+    preferences.sort_by(|(_, left), (_, right)| right.partial_cmp(left).unwrap_or(Ordering::Equal));
+
+    for (tag, _) in &preferences {
+        if let Ok(locale) = Locale::parse(tag) {
+            if supported.contains(&locale) {
+                return locale;
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/detect.rs]: Updated `locale_from_accept_language` to skip non-positive q-values before exact or language-only matching, and added regression coverage for `de;q=0,en-US;q=0.5` plus the rejected language-only fallback case.

File: crates/ars-i18n/src/detect.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+///
+/// Returns the best matching locale from `supported`, preferring exact matches,
+/// then language-only matches, then the first supported locale. When
+/// `supported` is empty, this falls back to `en-US`.
+#[must_use]
+pub fn locale_from_accept_language(accept_language: &str, supported: &[Locale]) -> Locale {
+    let mut preferences = accept_language
+        .split(',')
+        .filter_map(|part| {
+            let mut iter = part.trim().splitn(2, ";q=");
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/detect.rs]: Switched the parser to split on `;`, trim parameters, and read `q` case-insensitively, so valid forms like `en; q=0.8` and `Q=0.8` now work. Added focused tests for whitespace and parameter-name casing.

File: crates/ars-i18n/src/detect.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        .collect::<Vec<_>>();
+
+    preferences.sort_by(|(_, left), (_, right)| right.partial_cmp(left).unwrap_or(Ordering::Equal));
+
+    for (tag, quality) in &preferences {
+        if *quality <= 0.0 {
+            continue;
+        }
+
+        if let Ok(locale) = Locale::parse(tag) {
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/detect.rs]: Updated the matcher to treat `*` as a real language range and only apply it to supported locales that are not already covered by a more specific header range. That lets `en-US;q=0.4,*;q=0.8` select the wildcard-eligible default while still preventing cases like `de;q=0,*;q=0.8` from reselecting an explicitly rejected locale. I also added focused regression tests for both wildcard matching and wildcard exclusion, and synced the spec example in `04-internationalization.md` to the current parser behavior.

File: crates/ars-i18n/src/detect.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+                    .iter()
+                    .find(|supported_locale| supported_locale.language() == locale.language())
+                {
+                    return matched.clone();
+                }
+            }
+            PreferenceTag::Language(language) => {
+                if let Some(matched) = supported
+                    .iter()
+                    .find(|supported_locale| supported_locale.language() == *language)
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/detect.rs]: Updated both the plain language-range branch and the locale-to-language fallback branch to skip supported locales that have their own explicit locale range elsewhere in the header. That means cases like `en;q=0.9,en-US;q=0.1` and `en-CA;q=0.9,en-US;q=0.1` now select `en-GB` when it is available instead of letting the broader fallback grab the explicitly de-prioritized `en-US`. I added focused regression tests for both cases and reran `cargo test -p ars-i18n detect --quiet`, the annotated `cargo llvm-cov` slice for `detect.rs`, and a full `cargo xci` pass.

File: crates/ars-i18n/src/detect.rs [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        Some(PreferenceTag::Language(locale.language().to_string()))
+    } else {
+        Some(PreferenceTag::Locale(locale))
+    }
+}
+
+fn parse_quality(value: &str) -> f32 {
+    match value.parse::<f32>() {
+        Ok(quality) if quality.is_finite() => quality,
+        _ => 1.0,
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/detect.rs]: Tightened this to parse only RFC-style qvalues in the `0..=1` range. The parser now accepts `0`, `1`, `0.xxx`, and `1.000`-style forms with at most three fractional digits, and treats out-of-range or malformed values like `2`, `1.001`, and `0.1234` as invalid so they fall back to the default weight instead of outranking valid preferences. I also added regression tests for the out-of-range, invalid-precision, and boundary cases, and updated the `04-internationalization.md` example to match the implementation.

File: crates/ars-i18n/src/detect.rs [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+                }) {
+                    return matched.clone();
+                }
+            }
+            PreferenceTag::Locale(locale) => {
+                if supported.contains(locale) {
+                    return locale.clone();
+                }
+
+                if let Some(matched) = supported.iter().find(|supported_locale| {
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-i18n/src/detect.rs:102]: Updated locale-range matching so a parsed locale range now matches supported tags that are more specific on a subtag boundary before we drop to language-only fallback. That means a header like `zh-Hant` now correctly selects `zh-Hant-TW` instead of falling through to a generic `zh-*` match such as `zh-Hans-CN`. I also reused that same range check when excluding supported locales from broader language and wildcard matches, so `zh;q=0.9,zh-Hant;q=0.1` no longer lets the broad `zh` rule grab a locale already covered by the more-specific `zh-Hant` range. Added focused regressions for the direct locale-range match, the broader language exclusion, and the wildcard exclusion case, and synced the `04-internationalization.md` example to the implementation.
